v0.32 release notes (#4496)
PennyLaneAI/pennylane: doc/releases/changelog-0.32.0.md

<h3>New features since last release</h3>

<h4>Encode matrices using a linear combination of unitaries ⛓️️</h4>

> * It is now possible to encode an operator `A` into a quantum circuit by decomposing it into
>   a linear combination of unitaries and using the
>   [StatePrep](https://docs.pennylane.ai/en/stable/code/api/pennylane.StatePrep.html) and
>   [Select](https://docs.pennylane.ai/en/stable/code/api/pennylane.Select.html) operations to input
>   the coefficients and unitaries, respectively, into a quantum circuit.

Author: This is quite lengthy for an opening statement. How about:

```suggestion
* It is now possible to encode an operator `A` into a quantum circuit by decomposing it into
  a linear combination of unitaries using PREP 
  ([StatePrep](https://docs.pennylane.ai/en/stable/code/api/pennylane.StatePrep.html)) and
  SELECT ([Select](https://docs.pennylane.ai/en/stable/code/api/pennylane.Select.html)) 
  routines.
```

Author: Also, do the features require that `A` already be a Hamiltonian / LCU?

Reviewer: Yes for this release we require A to be self adjoint

  [(#4431)](https://github.com/PennyLaneAI/pennylane/pull/4431)
  [(#4437)](https://github.com/PennyLaneAI/pennylane/pull/4437)
  [(#4444)](https://github.com/PennyLaneAI/pennylane/pull/4444)
  [(#4450)](https://github.com/PennyLaneAI/pennylane/pull/4450)

  Consider an operator `A` composed of a linear combination of Pauli terms:

  ```pycon
  >>> A = qml.PauliX(2) + 2 * qml.PauliY(2) + 3 * qml.PauliZ(2)
  ```

  A decomposable block encoding circuit can be created:

  ```python
  def block_encode(A, control_wires):
      probs = A.coeffs / np.sum(A.coeffs)
      state = np.pad(np.sqrt(probs, dtype=complex), (0, 1))
      unitaries = A.ops

      qml.StatePrep(state, wires=control_wires)
      qml.Select(unitaries, control_wires=control_wires)
      qml.adjoint(qml.StatePrep)(state, wires=control_wires)
  ```

  ```pycon
  >>> print(qml.draw(block_encode, show_matrices=False)(A, control_wires=[0, 1]))
  0: ─╭|Ψ⟩─╭●──────────╭|Ψ⟩†─┤  
  1: ─╰|Ψ⟩─├●──────────╰|Ψ⟩†─┤  
  2: ──────╰Select(M0)───────┤  
  ```

  This circuit can be used as a building block within a larger QNode to perform algorithms such as

>   the [quantum singular value transformation](https://docs.pennylane.ai/en/stable/code/api/pennylane.QSVT.html).

Author: might be helpful to mention this application as well?

```suggestion
  the [quantum singular value transformation](https://docs.pennylane.ai/en/stable/code/api/pennylane.QSVT.html) and [Hamiltonian simulation](https://codebook.xanadu.ai/H.6).
```

  routines.
  [(#4431)](https://github.com/PennyLaneAI/pennylane/pull/4431)
  [(#4437)](https://github.com/PennyLaneAI/pennylane/pull/4437)
  [(#4444)](https://github.com/PennyLaneAI/pennylane/pull/4444)
  [(#4450)](https://github.com/PennyLaneAI/pennylane/pull/4450)

  Consider an operator `A` composed of a linear combination of Pauli terms:

  ```pycon
  >>> A = qml.PauliX(2) + 2 * qml.PauliY(2) + 3 * qml.PauliZ(2)
  ```

  A decomposable block-encoding circuit can be created:

  ```python
  def block_encode(A, control_wires):
      probs = A.coeffs / np.sum(A.coeffs)
      state = np.pad(np.sqrt(probs, dtype=complex), (0, 1))
      unitaries = A.ops

      qml.StatePrep(state, wires=control_wires)
      qml.Select(unitaries, control_wires=control_wires)
      qml.adjoint(qml.StatePrep)(state, wires=control_wires)
  ```

  ```pycon
  >>> print(qml.draw(block_encode, show_matrices=False)(A, control_wires=[0, 1]))
  0: ─╭|Ψ⟩─╭●──────────╭|Ψ⟩†─┤  
  1: ─╰|Ψ⟩─├●──────────╰|Ψ⟩†─┤  
  2: ──────╰Select(M0)───────┤  
  ```

  This circuit can be used as a building block within a larger QNode to perform algorithms such as

>   the [quantum singular value transformation](https://docs.pennylane.ai/en/stable/code/api/pennylane.QSVT.html) and [Hamiltonian simulation](https://codebook.xanadu.ai/H.6).

Reviewer: ```suggestion
  [QSVT](https://docs.pennylane.ai/en/stable/code/api/pennylane.QSVT.html) and [Hamiltonian simulation](https://codebook.xanadu.ai/H.6).
```

  [(#4395)](https://github.com/PennyLaneAI/pennylane/pull/4395)
  [(#4479)](https://github.com/PennyLaneAI/pennylane/pull/4479)
  [(#4493)](https://github.com/PennyLaneAI/pennylane/pull/4493)

  ```python
  def find_coeffs(p):
      mat = np.array([[3, p], [p, 3]])
      A = qml.pauli_decompose(mat)
      return A.coeffs
  ```

  ```pycon
  >>> import jax
  >>> from jax import numpy as np
  >>> jax.jacobian(find_coeffs)(np.array(2.))
  Array([0., 1.], dtype=float32, weak_type=True)
  ```
  
<h4>Monitor PennyLane's inner workings with logging 📃</h4>

* Python-native logging can now be enabled with `qml.logging.enable_logging()`.
  [(#4377)](https://github.com/PennyLaneAI/pennylane/pull/4377)
  [(#4383)](https://github.com/PennyLaneAI/pennylane/pull/4383)

  Consider the following code that is contained in `my_code.py`:

  ```python
  import pennylane as qml
  qml.logging.enable_logging()  # enables logging

  dev = qml.device("default.qubit", wires=2)

  @qml.qnode(dev)
  def f(x):
      qml.RX(x, wires=0)
      return qml.state()

  f(0.5)
  ``` 

  Executing `my_code.py` with logging enabled will detail every step in PennyLane's
  pipeline that gets used to run your code.

>   ```commandline

Reviewer: ```suggestion
  ```
```

  [logging docs page](https://docs.pennylane.ai/en/latest/introduction/logging.html) for more
  details!

<h4>More input states for quantum chemistry calculations ⚛️</h4>

* Input states obtained from advanced quantum chemistry calculations can be used in a circuit. 
  [(#4427)](https://github.com/PennyLaneAI/pennylane/pull/4427)
  [(#4433)](https://github.com/PennyLaneAI/pennylane/pull/4433)
  [(#4461)](https://github.com/PennyLaneAI/pennylane/pull/4461)
  [(#4476)](https://github.com/PennyLaneAI/pennylane/pull/4476)

  Quantum chemistry calculations rely on an initial state that is typically selected to be the
  trivial Hartree-Fock state. For molecules with a complicated electronic structure, using initial
  states obtained from affordable post-Hartree-Fock calculations helps to improve the efficiency of
  the quantum simulations. These calculations can be done with external quantum chemistry libraries
  such as `PySCF`.

  It is now possible to import a `PySCF` solver object in PennyLane and extract the corresponding
  wave function in the form of a state vector that can be directly used in a circuit. The user needs

>   to run the classical quantum chemistry calculations first and then use the `qml.import_state`

Reviewer 2: Correct to `qml.qchem.import_state`.

  >>> wf_cisd = qml.qchem.import_state(myci, tol=1e-1)
  >>> print(wf_cisd)
  [ 0.        +0.j  0.        +0.j  0.        +0.j  0.1066467 +0.j
    1.        +0.j  0.        +0.j  0.        +0.j  0.        +0.j
    2.        +0.j  0.        +0.j  0.        +0.j  0.        +0.j
   -0.99429698+0.j  0.        +0.j  0.        +0.j  0.        +0.j]
  ```

  The currently supported objects are RCISD, UCISD, RCCSD, and UCCSD which correspond to 
  restricted (R) and unrestricted (U) configuration interaction (CI) and coupled cluster (CC) 
  calculations with single and double (SD) excitations.

* `qml.import_state` is now accounted for in `doc/introduction/chemistry.rst`, adding the documentation for the function.
  [(#4461)](https://github.com/PennyLaneAI/pennylane/pull/4461)

* Input types and sources for external wavefunctions and operators for `qml.import_state` 

>   and `qml.import_operator` are clarified. [(#4476)](https://github.com/PennyLaneAI/pennylane/pull/4476)

Author: Should these be in the documentation section?

Reviewer: Yes this entry has been removed (except we keep the PR number)

  [(#4432)](https://github.com/PennyLaneAI/pennylane/pull/4432)

  Post-measurement reset can be activated by setting `reset=True` when calling
  [qml.measure](https://docs.pennylane.ai/en/stable/code/api/pennylane.measure.html).
  In this version of PennyLane, executing circuits with qubit reuse will result in the
  [defer_measurements](https://docs.pennylane.ai/en/latest/code/api/pennylane.defer_measurements.html)
  transform being applied. This transform replaces each reused wire with an *additional* qubit.
  However, future releases of PennyLane will explore device-level support for qubit reuse without
  consuming additional qubits.

  Qubit reuse and reset is also fully differentiable:

  ```python
  dev = qml.device("default.qubit", wires=4)

  @qml.qnode(dev)
  def circuit(p):
      qml.RX(p, wires=0)
      m = qml.measure(0, reset=True) 
      qml.cond(m, qml.Hadamard)(1)

      qml.RX(p, wires=0)
      m = qml.measure(0)
      qml.cond(m, qml.Hadamard)(1)
      return qml.expval(qml.PauliZ(1))
  ```

  ```pycon
  >>> jax.grad(circuit)(0.4)
  Array(-0.35867804, dtype=float32, weak_type=True)
  ```
  
  You can read more about mid-circuit measurements

>   [here](https://docs.pennylane.ai/en/latest/introduction/measurements.html#mid-circuit-measurements-and-conditional-operations),

Reviewer: ```suggestion
  [in our documentation](https://docs.pennylane.ai/en/latest/introduction/measurements.html#mid-circuit-measurements-and-conditional-operations),
```

    1.        +0.j  0.        +0.j  0.        +0.j  0.        +0.j
    2.        +0.j  0.        +0.j  0.        +0.j  0.        +0.j
   -0.99429698+0.j  0.        +0.j  0.        +0.j  0.        +0.j]
  ```

  The state vector can be implemented in a circuit using `qml.StatePrep`.

  ```pycon
  >>> dev = qml.device('default.qubit', wires=4)
  >>> @qml.qnode(dev)
  ... def circuit():
  ...     qml.StatePrep(wf_cisd, wires=range(4))
  ...     return qml.state()
  >>> print(circuit())
  [ 0.        +0.j  0.        +0.j  0.        +0.j  0.1066467 +0.j
    1.        +0.j  0.        +0.j  0.        +0.j  0.        +0.j
    2.        +0.j  0.        +0.j  0.        +0.j  0.        +0.j
   -0.99429698+0.j  0.        +0.j  0.        +0.j  0.        +0.j]
  ```

  The currently supported post-Hartree-Fock methods are RCISD, UCISD, RCCSD, and UCCSD which 
  denote restricted (R) and unrestricted (U) configuration interaction (CI) and coupled cluster (CC) 
  calculations with single and double (SD) excitations.

<h4>Reuse and reset qubits after mid-circuit measurements ♻️</h4>

* PennyLane now allows you to define circuits that reuse a qubit after a mid-circuit

>   measurement has taken place. Optionally, the wire can also be reset to the :math:`|0\rangle`

Reviewer 2: The math equation does not render correctly.

  [(#4301)](https://github.com/PennyLaneAI/pennylane/pull/4301)

* `qml.math.reduced_dm` has been removed. Use `qml.math.reduce_dm` or `qml.math.reduce_statevector` instead.
  [(#4301)](https://github.com/PennyLaneAI/pennylane/pull/4301)

* The `qml.specs` dictionary no longer supports direct key access to certain keys. 
  [(#4301)](https://github.com/PennyLaneAI/pennylane/pull/4301)

  Instead, these quantities can be accessed as fields of the new `Resources` object saved under
  `specs_dict["resources"]`:

  - `num_operations` is no longer supported, use `specs_dict["resources"].num_gates`
  - `num_used_wires` is no longer supported, use `specs_dict["resources"].num_wires`
  - `gate_types` is no longer supported, use `specs_dict["resources"].gate_types`
  - `gate_sizes` is no longer supported, use `specs_dict["resources"].gate_sizes`
  - `depth` is no longer supported, use `specs_dict["resources"].depth`

* `qml.math.purity`, `qml.math.vn_entropy`, `qml.math.mutual_info`, `qml.math.fidelity`,
  `qml.math.relative_entropy`, and `qml.math.max_entropy` no longer support state vectors as
  input.
  [(#4322)](https://github.com/PennyLaneAI/pennylane/pull/4322)

> * The Pauli-X-term in `qml.pulse.transmon_drive` has been removed in accordance with [1904.06560](https://arxiv.org/abs/1904.06560)

Author: Already mentioned elsewhere? (line 376)

  [(#4322)](https://github.com/PennyLaneAI/pennylane/pull/4322)

* `qml.ctrl(qml.PauliX)` returns a `CNOT`, `Toffoli`, or `MultiControlledX` operation instead of `Controlled(PauliX)`.
  [(#4339)](https://github.com/PennyLaneAI/pennylane/pull/4339)

* When given a callable, `qml.ctrl` now does its custom pre-processing on all queued operators from the callable.
  [(#4370)](https://github.com/PennyLaneAI/pennylane/pull/4370)

* The `qchem` functions `primitive_norm` and `contracted_norm` have been modified to be compatible with
  higher versions of SciPy. The private function `_fac2` for computing double factorials has also been added.
  [#4321](https://github.com/PennyLaneAI/pennylane/pull/4321)

* `tape_expand` now uses `Operator.decomposition` instead of `Operator.expand` in order to make
  more performant choices.
  [(#4355)](https://github.com/PennyLaneAI/pennylane/pull/4355)

* CI now runs tests with TensorFlow 2.13.0
  [(#4472)](https://github.com/PennyLaneAI/pennylane/pull/4472)

* All tests in CI and pre-commit hooks now enable linting.
  [(#4335)](https://github.com/PennyLaneAI/pennylane/pull/4335)

> * The default label for a `qml.StatePrep` operator is now `|Ψ⟩`.

Reviewer 3: ```suggestion
* The default label for a `StatePrepBase` operator is now `|Ψ⟩`.
```

  [(#4516)](https://github.com/PennyLaneAI/pennylane/pull/4516)

* The application of the `qml.defer_measurements` transform has been moved from 
  `QNode.construct` to `qml.Device.batch_transform` to allow more fine-grain 
  control over when `defer_measurements` should be used.
  [(#4432)](https://github.com/PennyLaneAI/pennylane/pull/4432)

* The label for `ParametrizedEvolution` can display parameters with the requested format as set by the 
  kwarg `decimals`. Array-like parameters are displayed in the same format as matrices and stored in the 
  cache.
  [(#4151)](https://github.com/PennyLaneAI/pennylane/pull/4151)

<h3>Breaking changes 💔</h3>

* Applying gradient transforms to broadcasted/batched tapes has been deactivated until it is consistently
  supported for QNodes as well.
  [(#4480)](https://github.com/PennyLaneAI/pennylane/pull/4480)

* Gradient transforms no longer implicitly cast `float32` parameters to `float64`. Finite difference differentiation
  with `float32` parameters may no longer give accurate results.
  [(#4415)](https://github.com/PennyLaneAI/pennylane/pull/4415)

* The `do_queue` keyword argument in `qml.operation.Operator` has been removed. Instead of

>   setting `do_queue=False`, use the `qml.QueuingManager.stop_recording()` context instead.

Reviewer 3: ```suggestion
  setting `do_queue=False`, use the `qml.QueuingManager.stop_recording()` context.
```
